File: src/quantification_processing.py

#Author: Franz AKE
#July 2022

#libs
import argparse
import pandas as pd

#Arguments Parser
parser = argparse.ArgumentParser(description='Processing of Salmon quantification file')
parser.add_argument('-salmon_quant', type=str, help='path of salmon quant file')
parser.add_argument('-output_file', type=str, help='path of output file')
args = parser.parse_args()

#Files
print("read file...")
all_samples = []
for filin in args.salmon_quant.split(","):
	#open file
	filetab = pd.read_csv(filin, sep="\t")
	filetab["Sample"] = filin
	all_samples.append(filetab)
qf = pd.concat(all_samples)

#expand gene name
name_tr = qf.Name.values[1]
if '|' in name_tr:
	qf = pd.concat([qf.Name.str.split("|", expand=True).loc[:,[0]], qf[['Length', 'NumReads','Sample']]], axis=1)
else:
	qf = qf.loc[:,['Name','Length','NumReads','Sample']]
qf.columns = ['transcript_id','length','NumReads','Sample']


#filter transcripts null
qf = qf.drop_duplicates()
grouped = qf.sort_values(['transcript_id','NumReads']).groupby("transcript_id")
out = list()
for tab in grouped:
	tab[1]['NumReads'] = tab[1]['NumReads'].mean()
	out.append(tab[1])
qf = pd.concat(out)
qf = qf[~qf.transcript_id.isin(qf[qf.NumReads==0].transcript_id)]
qf['Sample'] = 'averaged'
qf = qf.drop_duplicates()

#write csv and parquet file
qf.to_csv(args.output_file, sep="\t", doublequote=False, index=False)

File: src/quantification_processing.R

#libs
suppressWarnings(library(argparse, quietly=T, verbose=F))
suppressWarnings(library(dplyr, quietly=T, verbose=F))

#Argument parser
parser = ArgumentParser(description='Processing of salmon quantification_file')
parser$add_argument('genome', type="character", help='path of gtf file')
parser$add_argument('output_path', type="character", help='output path')
args = parser$parse_args()


#Opening

#GTF
message("GTF opening...")
gtf = rtracklayer::import(args$genome) %>%
    data.frame() %>%
    dplyr::filter(type=="exon")

message("GTF check the expected attributes to be present...")
if( FALSE %in% c(c("gene_id","transcript_id") %in% colnames(gtf)) ){ stop("Error! The attributes 'gene_id' or 'transcript_id' are expected in your GTF file") }

if( ("gene_name" %in% colnames(gtf)) & ("transcript_name" %in% colnames(gtf)) ) {
    gtf = gtf %>%
        dplyr::mutate(gene_name = ifelse(is.na(gene_name), gene_id, gene_name), transcript_name = ifelse(is.na(transcript_name), transcript_id, transcript_name)) %>%
        dplyr::distinct(seqnames, start, end, width, strand, gene_id, gene_name, transcript_id, transcript_name)
} else {
    gtf = gtf %>%
        dplyr::mutate(gene_name = gene_id, transcript_name = transcript_id) %>%
        dplyr::distinct(seqnames, start, end, width, strand, gene_id, gene_name, transcript_id, transcript_name)
}

#check1
if(nrow(gtf)==0){ stop("Error! GTF file contains no 'exon' entries ! Please provide valid GTF file") }

#SAMPLES QUANTIFICATION
message("Processing bulk quantification sample files...")
lapply(list.files(path = ".", pattern = "*.sf", full.names = T), function(x){
    #reading bulk counts
    curr = data.table::fread(x, nThread=1)
    curr$sample = x
    return(curr)
}) %>% data.table::rbindlist() %>%
    #get MeanTPM & discard null transcrips in bulk
    dplyr::group_by(Name) %>%
    dplyr::summarise(meanTPM = mean(TPM)) %>%
    dplyr::filter(meanTPM!=0) %>%
    dplyr::rename(transcript_id="Name") %>%
    #include gene metadata
    left_join(distinct(gtf, gene_name, transcript_name, transcript_id)) %>%
    #calculate Transcripts bulk % in gene
    dplyr::group_by(gene_name) %>%
    dplyr::reframe(transcript_name, bulk_TPMperc = meanTPM / sum(meanTPM)) %>%
    arrange(gene_name,transcript_name) -> bulk_quants

#writing merge quants
bulk_quants %>% stats::na.omit() %>% data.table::fwrite(file=args$output_path, nThread=1, row.names=F, sep="\t")

#filtering of GTF
gtf = dplyr::filter(gtf, gene_name %in% bulk_quants$gene_name) %>% stats::na.omit()

#check2
if(nrow(gtf)==0){ stop("Error! No gene in the GTF have been bulk quantified by Salmon quant! Please provide valid GTF file") }

#SPLIT GTF by chromosomes
message("GTF splitting...")
lapply(split(gtf, gtf$seqnames), function(x){
    #in case of GTF file from Ensembl

    # a. rename columns
    lookup = c(gene_type = "gene_biotype", transcript_type = "transcript_biotype")
    if("gene_biotype" %in% colnames(x)){
        rename(x, all_of(lookup))
    }

    # b. check gene version
    if("gene_version" %in% colnames(x)){
        x$gene_id = paste0(x$gene_id, ".", x$gene_version)
        x$transcript_id = paste0(x$transcript_id, ".", x$transcript_version)
    }

    #writing
    x = distinct(x, seqnames,start,end,width,strand,gene_name,transcript_name)
    if(nrow(x)!=0){ 
        data.table::fwrite(x, file = paste0(x$seqnames[1],".gtf"), sep="\t", nThread=1)
    }
}) -> writeds
